### PR TITLE
Add C++ MNIST example in benchmark

### PR DESCRIPTION
**Pull request by Yancey1989**

### CODE REVIEW

File: benchmark/go/mnist.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
+package main
```
Comment (Yancey1989) [inline benchmark/go/mnist.go:1]: This file is reduplicated with https://github.com/wangkuiyi/gotorch/blob/develop/data_test.go , I will polish it in another PR.

File: benchmark/cpp/main.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      loss_value = loss.template item<float>();
+      loss.backward();
+      optimizer.step();
+      if (batch_idx % 200 == 0) {
+        std::printf("Train Epoch: %ld, Batch: %zu, Loss: %.4f\n", epoch,
+                    batch_idx, loss_value);
+      }
+      batch_idx++;
+    }
+    float duration =
```
Comment (ljk53) [inline benchmark/cpp/main.cpp:83]: clock() doesn't seem to be accurate - let's try using std::chrono::high_resolution_clock:

```
#include <chrono>
using namespace std::chrono;
...
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
...
    high_resolution_clock::time_point t2 = high_resolution_clock::now();
    duration<double> time_span = duration_cast<duration<double>>(t2 - t1);
    int throughput = (train_dataset_size * 1.0 / time_span.count());
```

File: benchmark/README.md [hunk 3]
```diff
@@ -14,6 +14,13 @@ structure, and dataset, type the following command:
...
 ```
 
+To run the Go version of MNIST training with the same parameters above, type
+the following command:
+
+```bash
+cd benchmark && make && time ./mnist
+```
+
 Both programs use only CPU but no GPU.
```
Comment (wangkuiyi) [inline benchmark/README.md:24]: It is no longer "both", but "all".

File: benchmark/main.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+#include <chrono>  // NOLINT
+#include <cstddef>
+#include <cstdio>
+#include <iostream>
+#include <string>
+#include <vector>
+
+// Where to find the MNIST dataset.
+const char* kDataRoot = "/Users/yancey/.cache/mnist";
```
Comment (wangkuiyi) [inline benchmark/main.cpp:12]: It seems that this constant value is defined but not used.

File: benchmark/Makefile [hunk 5]
```diff
@@ -0,0 +1,16 @@
+all: main
+
+main:
+	clang++ -std=c++14 \
```
Comment (QiJune) [inline benchmark/Makefile:4]: Dose `-O3` flag contribute to the benchmark? We could have a try.

Comment (wangkuiyi) [reply benchmark/Makefile:4]: I agree. I think -O3 could be more useful when used to build `libcgotorch.so`.

Comment (Yancey1989) [reply benchmark/Makefile:4]: I got no significant improvement on my host, average the throughput of 5 times running.

Enable `-O3`
``` bash
➜  gotorch git:(mnist_benchmark) ✗ make -C benchmark
clang++ -std=c++14 -O3 \
	-I /Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/include \
	-I /Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/include/torch/csrc/api/include \
	-L /Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/lib \
	main.cpp \
	-o mnist \
	-Wl,-rpath,/Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/lib \
	-pthread \
	-lc10 -ltorch -ltorch_cpu \
	-D_GLIBCXX_USE_CXX11_ABI=0
➜  gotorch git:(mnist_benchmark) ✗ for i in `seq 5`; do ./benchmark/mnist; done | awk '{sum+=$3} END {print sum/NR}'
18927.3
```

Without `-O3`

``` bash
➜  gotorch git:(mnist_benchmark) make -C benchmark
clang++ -std=c++14 \
	-I /Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/include \
	-I /Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/include/torch/csrc/api/include \
	-L /Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/lib \
	main.cpp \
	-o mnist \
	-Wl,-rpath,/Users/yancey/go/src/github.com/wangkuiyi/gotorch/cgotorch/libtorch/lib \
	-pthread \
	-lc10 -ltorch -ltorch_cpu \
	-D_GLIBCXX_USE_CXX11_ABI=0
➜  gotorch git:(mnist_benchmark) for i in `seq 5`; do ./benchmark/mnist; done | awk '{sum+=$3} END {print sum/NR}'
19361.8
```

**Pull request merged**